server/controllers/proxyController.js

```diff
@@ -16,29 +17,63 @@ const proxy = httpProxy.createProxyServer({
 proxy.on('proxyReq', function(proxyReq, req, res, options) {
   proxyReq.setHeader('authorization', `Bearer ${AIRTABLE_API_KEY}`);
 });
-
 proxy.on('proxyRes', function(proxyRes, req, res) {
+  var body = [];
   proxyRes.on('data', function(chunk) {
+    body.push(chunk);
+  });
+  proxyRes.on('end', function() {
+    const buffer = Buffer.concat(body);
+    let proxyPayload = null;
+
     if (proxyRes.headers[`${CONTENT_ENCODING}`] == `${GZIP}`) {
-      zlib.gunzip(chunk, function(err, dezipped) {
-        let unzipPayload = dezipped.toString();
-        cache.set(req, JSON.parse(unzipPayload));
-        return res.send(JSON.parse(unzipPayload));
+      zlib.gunzip(buffer, function(err, decoded) {
+        if (!err) {
+          proxyPayload = JSON.parse(decoded.toString());
+          if (req.method === 'GET') {
```

Comment:
We can DRY these lines up! One approach is to break out the gzip/deflate logic separately:

```javascript
async function handlePayload(buffer, encoding) {
  return new Promise((resolve, reject) => {
     if (encoding === GZIP) {
        zlib.gunzip(buffer, function(err, decoded) {
           if (err) return reject(err);
           resolve(decoded.toString());
        });
     } else if (encoding === DEFLATE) {
       zlib.inflate(buffer, function(err, decoded) {
           if (err) return reject(err);
           resolve(decoded.toString());
       });
     } else {
        resolve(buffer.toString());
     }
  });
}
```

...and then perform the cache setting and response handling after!

```javascript
const proxyPayload = await handlePayload(buffer);
if (req.method === 'GET') {
    cache.set(req, proxyPayload);
}
res.status(proxyRes.statusCode).send(proxyPayload);
res.end();
```

Reply:
@ethanlee16 beautiful, can't cry bc my tears are all DRY